### DISCUSSION TITLE
Bio Site Form: Uncontrolled vs. Controlled Variable

### DISCUSSION THREAD

**Discussion by cmeffley** [nss-evening-cohort-14/evening-client-side#245] [Help Me!] [answered]
Chris Meffley

No other tickets that I see

Trying to edit my projects and unable to submit the form, getting the following error:
 Warning: A component is changing a controlled input to be uncontrolled. This is likely caused by the value changing from a defined to undefined, which should not happen. Decide between using a controlled or uncontrolled input element for the lifetime of the component.

https://reactjs.org/docs/uncontrolled-components.html#:~:text=In%20a%20controlled%20component%2C%20form,form%20values%20from%20the%20DOM.
https://stackoverflow.com/questions/47012169/a-component-is-changing-an-uncontrolled-input-of-type-text-to-be-controlled-erro
https://stackoverflow.com/questions/51406927/reactjs-warning-a-component-is-changing-an-uncontrolled-input-of-type-text-to

I'm trying to make essentially a mirror of what a non admin user can see, but with editing abilities and I've looked into the useRef for uncontrolled, but unfamiliar with it, and feel like I'm just missing one or two things.
I've referenced other projects and everything looks similar
I've `console.warn` and I am getting all my projects and icon data where they need to be.

Edit Form:
https://github.com/cmeffley/personal-bio-site/blob/main/src/components/EditProjectsForm.js
Edit View:
https://github.com/cmeffley/personal-bio-site/blob/main/src/views/EditProjects.js
Promise Calls:
https://github.com/cmeffley/personal-bio-site/blob/main/src/helpers/data/ProjectData.js

**Comment by cmeffley** [comment, marked as answer]
I had to correct a few things:
- My form had `onChange={handleSubmit}`, but needed `onSubmit={handleSubmit}`, because the handleSubmit function was running each time I was changing the input value.
- When I did submit the form, I was setting the incorrect State. I was trying to set the edit State instead of the project State:
`updateProject(editProject, admin).then((project) => setChangeProjects(project));`  <---Correct
`updateProject(editProject, admin).then((project) => setEditProject(project));`  <---Incorrect
I had to bring in setChangeProjects in as a prop for it to be used and have State update.

Also, when I created my delete button, it had `type='submit'` which made it delete, but also create the project, because my addProject function was in the if else statement with the update project submit.

**Reply by drteresavasquez** [reply]
You are learning, Chris!!! Excellent!
